open_lm/file_utils.py

```diff
@@ -175,7 +181,76 @@ def source_exhausted(paths, shard_list_per_source):
     return False
 
 
-def get_string_for_epoch(num_samples, starting_chunk, paths, weights, min_shards_needed):
+def adjust_samples(shard_list, manifest, starting_index, num_workers, world_size):
+    """Adjust the number of samples to train on.
+
+    This function returns the number of samples that we will train on, so that no worker will keep going
+    while the others have stopped. This is calculated via the same assignment of nodes that wds.split_by_node
+    and wds.split_by_worker does.
+    """
+    num_samples = [manifest[j]["num_chunks"] for j in range(starting_index, starting_index + len(shard_list))]
+    samples_per_worker = [0 for _ in range(num_workers * world_size)]
+    for gpu in range(world_size):
+        samples_gpu = []
+        for n in islice(num_samples, gpu, None, world_size):
+            samples_gpu.append(n)
+
+        for worker in range(num_workers):
+            for s in islice(samples_gpu, worker, None, num_workers):
+                samples_per_worker[gpu * num_workers + worker] += s
+
+    return min(samples_per_worker) * num_workers * world_size
+
+
+def count_checkpoints(total_steps, args):
```

Comment:
This doesn't return anything, should we rename to "log_num_checkpoints()"?

Reply:
Got it, I'll rename this.